src/main/java/com/puppycrawl/tools/checkstyle/utils/ModuleReflectionUtils.java

```diff
@@ -0,0 +1,158 @@
+     * @param packageName the package name to use
+     * @param loader the class loader used to load Checkstyle package name
+     * @return the set of checkstyle's module classes
+     * @throws IOException if the attempt to read class path resources failed
+     * @see #isCheckstyleModule(Class)
+     */
+    public static Set<Class<?>> getCheckstyleModulesRecursive(
+            String packageName, ClassLoader loader) throws IOException {
+        final ClassPath classPath = ClassPath.from(loader);
+        return classPath.getTopLevelClassesRecursive(packageName).stream()
+                .map(ClassPath.ClassInfo::load)
+                .filter(ModuleReflectionUtils::isCheckstyleModule)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Gets checkstyle's modules (directly, not recursively) in the given packages.
+     * @param packages the collection of package names to use
+     * @param loader the class loader used to load Checkstyle package names
+     * @return the set of checkstyle's module classes
+     * @throws IOException if the attempt to read class path resources failed
+     * @see #isCheckstyleModule(Class)
+     */
+    public static Set<Class<?>> getCheckstyleModules(
+            Collection<String> packages, ClassLoader loader) throws IOException {
+        final ClassPath classPath = ClassPath.from(loader);
+        return packages.stream()
+                .flatMap(pkg -> classPath.getTopLevelClasses(pkg).stream())
+                .map(ClassPath.ClassInfo::load)
+                .filter(ModuleReflectionUtils::isCheckstyleModule)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Checks whether a class may be considered as a checkstyle module. Checkstyle's modules are
+     * non-abstract classes, which names do not start with the word 'Input' (are not input files for
+     * UTs), and are either checkstyle's checks, file sets, filters, file filters, or root module.
+     * @param clazz class to check.
+     * @return true if the class may be considered as the checkstyle module.
+     */
+    public static boolean isCheckstyleModule(Class<?> clazz) {
+        return isValidCheckstyleClass(clazz)
+            && (isCheckstyleCheck(clazz)
+                    || isFileSetModule(clazz)
+                    || isFilterModule(clazz)
+                    || isFileFilterModule(clazz)
+                    || isRootModule(clazz));
+    }
+
+    /**
+     * Checks whether a class extends 'AutomaticBean', is non-abstract, and doesn't start with the
+     * word 'Input' (are not input files for UTs).
+     * @param clazz class to check.
+     * @return true if a class may be considered a valid production class.
+     */
+    public static boolean isValidCheckstyleClass(Class<?> clazz) {
+        return AutomaticBean.class.isAssignableFrom(clazz)
+                && !Modifier.isAbstract(clazz.getModifiers())
+                && !clazz.getSimpleName().startsWith("Input");
```

Comment:
What problem with have with "Input"s ?
Main code should not be affected by test code. If we have some extra classes that are picked up we could rename them to be our of radar, or ignore them if them do not cause problems.
"test" code should not affect "main" code.